src/metal.cs

```diff
@@ -4164,24 +4554,609 @@ interface MTLCounterSampleBuffer {
+		void BuildAccelerationStructure (IMTLAccelerationStructure accelerationStructure, MTLAccelerationStructureDescriptor descriptor, IMTLBuffer scratchBuffer, nuint scratchBufferOffset);
+
+		[Abstract]
+		[Export ("refitAccelerationStructure:descriptor:destination:scratchBuffer:scratchBufferOffset:")]
+		void RefitAccelerationStructure (IMTLAccelerationStructure sourceAccelerationStructure, MTLAccelerationStructureDescriptor descriptor, [NullAllowed] IMTLAccelerationStructure destinationAccelerationStructure, IMTLBuffer scratchBuffer, nuint scratchBufferOffset);
+
+		[Abstract]
+		[Export ("copyAccelerationStructure:toAccelerationStructure:")]
+		void CopyAccelerationStructure (IMTLAccelerationStructure sourceAccelerationStructure, IMTLAccelerationStructure destinationAccelerationStructure);
+
+		[Abstract]
+		[Export ("writeCompactedAccelerationStructureSize:toBuffer:offset:")]
+		void WriteCompactedAccelerationStructureSize (IMTLAccelerationStructure accelerationStructure, IMTLBuffer buffer, nuint offset);
+
+		[Abstract]
+		[Export ("copyAndCompactAccelerationStructure:toAccelerationStructure:")]
+		void CopyAndCompactAccelerationStructure (IMTLAccelerationStructure sourceAccelerationStructure, IMTLAccelerationStructure destinationAccelerationStructure);
+
+		[Abstract]
+		[Export ("updateFence:")]
+		void UpdateFence (IMTLFence fence);
+
+		[Abstract]
+		[Export ("waitForFence:")]
+		void WaitForFence (IMTLFence fence);
+
+		[Abstract]
+		[Export ("useResource:usage:")]
+		void UseResource (IMTLResource resource, MTLResourceUsage usage);
+
+		[Abstract]
+		[Export ("useResources:count:usage:")]
+		void UseResources (IMTLResource[] resources, nuint count, MTLResourceUsage usage);
+
+		[Abstract]
+		[Export ("useHeap:")]
+		void UseHeap (IMTLHeap heap);
+
+		[Abstract]
+		[Export ("useHeaps:count:")]
+		void UseHeaps (IMTLHeap[] heaps, nuint count);
+
+		[Mac (11,0), iOS (14,0)]
+		[Abstract]
+		[Export ("sampleCountersInBuffer:atSampleIndex:withBarrier:")]
+		void SampleCountersInBuffer (MTLCounterSampleBuffer sampleBuffer, nuint sampleIndex, bool barrier);
+	}
+
+	interface IMTLVisibleFunctionTable {}
+
+	[Mac (11,0), iOS (14,0), NoTV]
+	[Protocol]
+	interface MTLVisibleFunctionTable : MTLResource {
+		[Abstract]
+		[Export ("setFunction:atIndex:")]
+		void SetFunction ([NullAllowed] IMTLFunctionHandle function, nuint index);
+
+		[Abstract]
+		[Export ("setFunctions:withRange:")]
+		void SetFunctions (IMTLFunctionHandle[] functions, NSRange range);
+	}
+
+	interface IMTLIntersectionFunctionTable {}
+
+	[Mac (11,0), iOS (14,0), NoTV]
+	[Protocol]
+	interface MTLIntersectionFunctionTable : MTLResource {
+		[Abstract]
+		[Export ("setBuffer:offset:atIndex:")]
+		void SetBuffer ([NullAllowed] IMTLBuffer buffer, nuint offset, nuint index);
+
+		[Abstract]
+		[Export ("setBuffers:offsets:withRange:")]
+		void SetBuffers (IMTLBuffer[] buffers, /* nuint[]*/ IntPtr offsets, NSRange range);
```

Comment:
Since consumers aren't supposed to implement this protocol themselves, it would make sense to implement a manual extension method on `IMTLIntersectionFunctionTable` that takes a `nuint[]` and calls this method.

Reply:
This is still not done as far as I can see.

Reply:
Looking at the header, the first argument is a C-style array as well, and as such must be bound using an IntPtr.

Unfortunately that's the case for the other types with a SetBuffers method as well:

https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443134-setbuffers: https://github.com/xamarin/xamarin-macios/blob/37a6569aee151e6603a4edb396b724cb8b0cbf48/src/metal.cs#L448-L449

https://developer.apple.com/documentation/metal/mtlintersectionfunctiontable/3578280-setbuffers?language=objc: https://github.com/xamarin/xamarin-macios/blob/37a6569aee151e6603a4edb396b724cb8b0cbf48/src/metal.cs#L3591-L3592

So we have existing broken API 😒 but let's make sure any new API we add is correct.